dsimporter.1.md
%dsimporter(1) dataset user manual | version 2.3.1 9a3d898
% R. S. Doiel and Tom Morrell
% 2025-07-10

# NAME

dsimporter

# SYNOPSIS

dsimporter [OPTIONS] C_NAME CSV_FILENAME KEY_COLUMN

# DESCRIPTION

__dsimporter__ is a tool to import CSV content into a dataset collection
where the column headings become the attribute names and the row values
become the attribute values.

# PARAMETERS

C_NAME
: If harvesting the dataset collection name to harvest the records to.

CSV_FILENAME
: The name of the CSV file to import

KEY_COLUMN
: The column name to use the they object key. If none is provided then
the first column is used as the object key. Keys values must be unique.


# OPTIONS

-help
: display help

-license
: display license

-version
: display version

-comma
: Set column delimiter

-comment
: Set row comment delimiter

-overwrite
: Overwrite objects on key collision

# EXAMPLES

Import a file with three columns

- item_code
- title
- location

The "item_code" is unique for each row. The data is stored
in a file called "books.csv". We are importing the CSV file
into a collections called. "shelves.ds"

~~~
dsimporter shelves.ds books.csv "item_code"
~~~




CHANGES.md
Release 2.3.2

Frame, clone, sample, sync and join support removed. The dsimporter cli removed (use jsonl dump and load instead). Fixed a bug in using the JSON API with GET where command line parameters were not getting processed. Also added support for standard encoding of web forms in queries.

Release 2.3.1:

Documentation and tests have been removed for frames, clone, and sample. Frame support was removed from datasetd.

Release 2.3.0:

This release fixed problems related to how a collection might incorrectly return a unknown type. v2.3.x cycle will depreciated features in preparation for v2.4.

Release 2.0.0-b06:

This release is a rewrite of version 1 focusing on removing features, clearifying useful concepts and abstracting the storage engines cleanly. The latter was done to allow the web implementation of dataset to achieve an appropriate performance and be able to scale to a larger number of collections and size of collections.


The dataset collection's structure has changed.

- a dataset is a directory containing a collection.json and codemeta.json file
- the collection.json no longer contains general metadata or maps to the keys and pairtree, it focuses on operational settings (e.g. storage type and access information)
    - when running dataset as a web service or on a shared user machine you can setup the database connection through the environment. I.e. set DATASET_DSN_URI value (DSN URI is formed with a protocol named for the SQL driver, a "://" and the DSN for that driver, e.g. "mysql://DB_USER:DB_PASSWD@/DB_NAME")
- a codemeta.json file is now used for holding general collection level metadata. [codemeta](https://codemeta.github.io/) has been adopted by the data science community for describing data and software
- additional JSON configuration files may be used to manage the collection dependent on storage engine
- Experimental features in dataset v2
    - Using SQL JSON columns to store our JSON documents in a relational database
        - SQLite 3
        - MySQL 8
        - Posgres 14.5 (very experimental, still learning Postgres)
    - Object level versioning 
    - Attachment versioning
    - dataset package to support Async, multi-process access via JSON stored in SQL databases


Golang package changes:

- Minimum Go version is now 1.19
- The dataset v2 package has been substantially reorganized and simplified, most things have changed
    - Collection.Init() now takes two parameters, collection name, an an optional DSN URI, if a DSN URI is provided it'll define the storage engine, e.g. a dataset.SQLSTORE)
    - Collection.DocPath() removed, doesn't make sense anymore since JSON may be stored in a SQL table
    - Collection.Read() only takes two parameters, not three
    - Collection.Keys() returns a list of keys and an error value
    - Collection.KeyExists() was renamed Collection.HasKey() to be more idiomatic in Go
    - Collection.FrameExists() was renamed Collection.HasFrame() to be more idiomatic in Go
    - Collection.Length() returns an int64 rather than an int
    - Collection.MetadataJSON() renamed Collection.Metadata() returns the codemeta JSON for the collection
    - Collection.UpdateMetatada() has been added, takes the name of a codemeta.json to replace the existing codemeta content for the collection.


libdataset:

- The C shared library implementation has been dropped for now do to the challenges of easily cross compiling releases

CLI changes:

- options have be restructured so that most come after the verb
- help has been restructured to better support focusing the help text on the task needed
- the command line version is single user, single process and be default assumes pairtree storage. 
- Experimental features
    - JSON stored in SQL database via JSON column types
        - SQLite3
        - MySQL 8
        - Postgres 14.5


Web Service changes:

- the web service is generally RESTful so the end points no longer map directly to the client syntax
- it is recommended to use SQL storage for your dataset collections explosed using the web service (e.q. SQLite3 or MySQL). This is because a pairtree collection doesn't provide docuemnt locking
- relies on SQL JSON columns for storage
- Access to the SQL storage engine is through either the environment or a URI expressing a storage type as protocol and a data source name to making the connection


Release 1.1.0:

Added attachment support for __datasetd__.

Updated the metadata fields to include richer PersonOrOrg data structures
for author, contributor, funder as well as added an annotation map field for custom metadata.

Added "MetadataJSON()" function for Collection to quickly copy out
the metadata values from a collection.

```
    c, err := dataset.Open("MyData.ds")
    ...
    defer c.Close()
    fmt.Printf("%s", c.MetadataJSON())
```

Added "MetadataUpdate()" function to update a collection's metadata.

```
    c, err := dataset.Open("MyData.ds")
    ...
    defer c.Close()
    meta := new(Collection)
    meta.Description = "A test dataset"
    meta.Version = "1.0.0"
    meta.Author = []*PersonOrOrg{
        &Person{
            Type: "Person",
            GivenName: "Jane",
            FamilyName: "Doe",
        },
    }
    err = c.MetadataUpdate(meta)
    ...
```

Depreciated dependency on namaste package and Namaste support in command line tools. Removed "collections.go and collections_test.go" from repository (redundant code). Updated libdataset/libdataset.go to hold functions that were needed for the C-Shared library from collections.go. The Namaste fields in the collection's metadata are now depreciated.

The dataset.Init() now places a lock file in the collection directory and leaves the collection in an "Open" state, it should be explicitly closed after Init is called.

E.g. 

```
   c, err := dataset.Init("MyData.ds")
   ...
   defer c.Close()
```

Removed "set_*" for collection metadata fields from libdataset.go. These should be set using the dataset command line tool only.


cmd/dsimporter/dsimporter.go
// dsimport.go is a command line program for working an dataset collections. It
// is using to import data from a CSV file into a collection.
//
// @author R. S. Doiel, <rsdoiel@caltech.edu>
// @author Tom Morrell, <tmorrell@caltech.edu>
//
// Copyright (c) 2023, Caltech
// All rights not granted herein are expressly reserved by Caltech.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice,
// this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice,
// this list of conditions and the following disclaimer in the documentation
// and/or other materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors
// may be used to endorse or promote products derived from this software without
// specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.
package main

import (
	"flag"
	"fmt"
	"os"
	"path"

	// Caltech Library packages
	"github.com/caltechlibrary/dataset/v2"
)

func main() {
	appName := path.Base(os.Args[0])
	// NOTE: The following will be set when version.go is generated
	version := dataset.Version
	releaseDate := dataset.ReleaseDate
	releaseHash := dataset.ReleaseHash
	fmtHelp := dataset.FmtHelp
	overwrite, comma, comment := false, "", ""
	lazyQuotes, trimSpace := false, false
	dsqueryHelpText, helpText := dataset.DSQueryHelpText, dataset.DSImporterHelpText
	datasetdHelpText, apiText, serviceText, yamlText := dataset.DatasetdHelpText, dataset.DatasetdApiText, dataset.DatasetdServiceText, dataset.DatasetdYAMLText
	datasetHelpText := dataset.DatasetHelpText
	
	showHelp, showVersion, showLicense := false, false, false
	flag.BoolVar(&showHelp, "help", false, "display help")
	flag.BoolVar(&showVersion, "version", false, "display version")
	flag.BoolVar(&showLicense, "license", false, "display license")
	flag.BoolVar(&overwrite, "overwrite", overwrite, "overwrite object on key collision")
	flag.StringVar(&comma, "comma", comma, "set column delimiter")
	flag.StringVar(&comment, "comment", comment, "set row comment delimiter")
	flag.BoolVar(&lazyQuotes, "lazy", lazyQuotes, "use lazy quotes")
	flag.BoolVar(&trimSpace, "trim", trimSpace, "trim leading space")
	flag.Parse()
	args := flag.Args()

	in := os.Stdin
	out := os.Stdout
	eout := os.Stderr

	if showHelp {
		//NOTE: handle help for topic request
		topic := "help"
		if (len(args) > 0) {
			topic = args[0];
		}
		switch topic {
		case "dsquery":
			fmt.Fprintf(out, "%s\n", fmtHelp(dsqueryHelpText, "dsquery", version, releaseDate, releaseHash))
		case "query":
			fmt.Fprintf(out, "%s\n", fmtHelp(dsqueryHelpText, "dsquery", version, releaseDate, releaseHash))
		case "dataset":
			fmt.Fprintf(out, "%s\n", fmtHelp(datasetHelpText, "datasetd", version, releaseDate, releaseHash))
		case "datasetd":
			fmt.Fprintf(out, "%s\n", fmtHelp(datasetdHelpText, "datasetd", version, releaseDate, releaseHash))
		case "api":
			fmt.Fprintf(out, "%s\n", fmtHelp(apiText, "datasetd", version, releaseDate, releaseHash))	
		case "service":
			fmt.Fprintf(out, "%s\n", fmtHelp(serviceText, "datasetd", version, releaseDate, releaseHash))	
		case "yaml":
			fmt.Fprintf(out, "%s\n", fmtHelp(yamlText, "datasetd", version, releaseDate, releaseHash))	
		case "config":
			fmt.Fprintf(out, "%s\n", fmtHelp(yamlText, "datasetd", version, releaseDate, releaseHash))	
		default:
			fmt.Fprintf(os.Stdout, "%s\n", fmtHelp(helpText, appName, version, releaseDate, releaseHash))
		}
		os.Exit(0)
	}
	if showVersion {
		fmt.Fprintf(os.Stdout, "%s %s %s\n", appName, version, releaseHash)
		os.Exit(0)
	}
	if showLicense {
		fmt.Fprintf(os.Stdout, "%s\n", dataset.LicenseText)
		os.Exit(0)
	}
	if len(args) != 3 {
		fmt.Fprintf(os.Stderr, "usage: %s C_NAME CSV_FILENAME KEY_COLUMN", appName)
		os.Exit(10)
	}

	// Setup our POSIX IO options
	var err error

	// Create a DSQuery object and evaluate the command line options
	app := new(dataset.DSImport)
	app.Overwrite = overwrite
	app.LazyQuotes = lazyQuotes
	app.TrimLeadingSpace = trimSpace
	if comma != "" {
		app.Comma = comma
	}
	if comment != "" {
		app.Comment = comment
	}
	cName, csvName, keyColumn := args[0], args[1], args[2]

	// Handle arranging our input data
	if csvName != "" && csvName != "-" {
		in, err = os.Open(csvName)
		if err != nil {
			fmt.Fprintf(os.Stderr, "%s", err)
			os.Exit(64)
		}
		defer in.Close()
	}
	if err := app.Run(in, out, eout, cName, keyColumn); err != nil {
		fmt.Fprintf(os.Stderr, "%s\n", err)
		os.Exit(10)
	}
}
